### PR TITLE
Autolink when edititng comments except wehn explicit link removal

### PR DESCRIPTION
**Pull request by markarmanus**
### Details

The issue is when we edit a comment that has links we dont autolink them so that you can click on them. This is because we purposely removed autolnking in this PR https://github.com/Expensify/App/issues/9090
The reason we removed it was when a user would delete the link from a comment it would re genrate it again.

The solution is this PR is to always generate a link unles a user explicitly removes it while editing. This requires comapring the original comment and after deleting comment, detecting which links if any the user removed explicitly and delete them.

### Fixed Issues

$ https://github.com/Expensify/App/issues/13221 
$ https://github.com/Expensify/App/issues/13221#issuecomment-1338132072

### Tests

Same as QA

- [x] Verify that no errors appear in the JS console

### Offline tests

No specefic difference on offline. should behave exactly the same as online.
### QA Steps

1. Write a new message to any chat. 
       (e.g from `` to `hello`)
2. edit the message and add a link 
       (e.g from `hello`  to `hello www.google.com`)
3. confirm the link is generated and can be clicked on.
4. edit message again now add more normal text 
       (e.g  `hello [www.google.com](https://www.google.com)` to.  `hello [www.google.com](https://www.google.com) bye`)
5. confirm the link is still there
6. edit message again and remove the link part and the [] 
       (e.g from `hello [www.google.com](www.google.com) bye` to `hello www.google.com bye`)
7. confirm the link is removed 
8. edit the message again and any text 
       (e.g from `hello www.google.com bye` to `hello www.google.com bye newText`)
9. confirm the link is generated again. 
10. delete only the () part of the link
       (e.g from `hello [www.google.com](https://www.google.com) bye newText` to `hello [www.google.com] newText bye`)
11. confirm the link is removed but the [] are still there. (i.e comment should now be `hello [www.google.com] newText bye`
12. add a second link to the message
       (e.g from `hello [www.google.com] bye newText` to `hello [www.google.com] bye newText facebook.com`)
13. confirm two links are generated one for google and one for facebook. and the google one is in []
14. now delete the () part of the facebook link 
       (e.g from `helle [[www.google.com](https://www.google.com)] bye newText [facebook.com](https://www.facebook.com)` to `helle [[www.google.com](https://www.google.com)] bye newText facebook.com`
15. confirm only the facebook link is deleted.

Note: if the comment has the same link multiple times it could cause odd behaviour but its an edge case so should be okay.
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.
off every checkbox in the PR author checklist, including those that don't apply to this PR.

Desktop,Android and IOS

https://user-images.githubusercontent.com/33489215/207224792-074ed567-7cd4-4088-9488-2e3c86f0773b.mp4

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/33489215/207952329-39b988d2-4e89-4871-b9bd-6ef778884098.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/33489215/207952261-6ad7d3fd-0a80-4ba2-91cb-44930092c038.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/33489215/207952198-8636c194-1794-4bd6-b958-4b702b1e4353.mov

</details>

<details>
<summary>Desktop</summary>

 Please refer to video above with all Desktop, Android and IOS
</details>

<details>
<summary>iOS</summary>
 Please refer to video above with all Desktop, Android and IOS

</details>

<details>
<summary>Android</summary>
 Please refer to video above with all Desktop, Android and IOS

</details>

### CODE REVIEW

**Comment by eVoloshchak** [comment]
The code looks good and works well, there are a couple of comments that need to be changed

@markarmanus, as per [this](https://github.com/Expensify/App/issues/13221#issuecomment-1340118644) discussion, could you please add tests for all the cases you tested in the video?

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:896]:
```suggestion
 * Extracts all links in a markdown comment and returns them in a list.
```

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:904]: Could you add a little more detailed comment, why exactly are we picking the first element of `match`?

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:909]:
```suggestion
 * Compares two markdown comments and returns a list of the links removed in a new comment.
```

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:922]:
```suggestion
 * Removes the links in a markdown comment.
```

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:933]: As per [STYLE.MD](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#object--array-methods) use `_.each`

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:953]:
```suggestion
```

File: src/libs/actions/Report.js
Comment (eVoloshchak) [inline src/libs/actions/Report.js:956]:
```suggestion
    return removeLinks(markdownWithAutoLinks, removedLinks);
```

**Review by eVoloshchak — CHANGES_REQUESTED** [review]

**Comment by cead22** [comment]
Yes, please add unit tests for all the cases you tested manually 🙇

**Comment by markarmanus** [comment]
@cead22 @eVoloshchak Added unit tests. Also for some reason re requesting the review from @eVoloshchak removed @cead22. apologies for that. i dont seem to have premissions to add you again.

**Comment by markarmanus** [comment]
@cead22 @eVoloshchak any updates ?

**Comment by eVoloshchak** [comment]
Reviewing today

**Comment by markarmanus** [comment]
@eVoloshchak @cead22 Sorry to be buggy, but i only have sometime to work on this today and i dont want to miss the 3 days bonus :D

File: tests/actions/ReportTest.js
Comment (cead22) [inline tests/actions/ReportTest.js:417]:
```suggestion
        /* This tests a variety of scenarios when a user edits a comment.
         * We should generate a link when editing a message unless the link was
         * already in the comment and the user deleted it on purpose.
         */
```

**Review by cead22 — CHANGES_REQUESTED** [review]
Sorry, but there's a few items left to polish this PR

**Comment by markarmanus** [comment]
@cead22  Done andNo worries at all, thats the entire point of PR reviews. to make sure the code quality is up to snuff. You have a good eye.

**Comment by eVoloshchak** [comment]
@markarmanus, there are a couple of issues left with the PR description 

1. Change the PR title to be more informative, it should reflect what has been done/fixed
2. Upload a separate video for each platform, following the PR template

**Comment by eVoloshchak** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/9059945/207822274-c9f2627d-6e5b-4a55-8115-c2c24954d415.mov




</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/9059945/207822361-2ea024cf-95cd-4d44-9002-1170b48d46a3.mov




</details>

<details>
<summary>Android</summary>


https://user-images.githubusercontent.com/9059945/207822555-e7053075-3aaf-442c-92e9-a814c9e16952.mp4

-- add screenshots or videos here -->

</details>


**Note:** I couldn't test `web` and `mWeb` due to the [issue](https://expensify.slack.com/archives/C01GTK53T8Q/p1670943366932199) of not being able to sign in.

**Review by cead22 — DISMISSED** [review]
>@cead22 Done and no worries at all, thats the entire point of PR reviews. to make sure the code quality is up to snuff. You have a good eye.

Thanks for understanding

**Comment by markarmanus** [comment]
@eVoloshchak I cant seem to run the local web due to the issue you mentioend. are you aware of any work aroundts at this point ?

**Comment by eVoloshchak** [comment]
> @eVoloshchak I cant seem to run the local web due to the issue you mentioend. are you aware of any work aroundts at this point ?

As per [this](https://expensify.slack.com/archives/C01GTK53T8Q/p1670929674365129?thread_ts=1670706653.367089&cid=C01GTK53T8Q) comment

>I would say if you can test on Desktop, I would just use that for web screenshots and note this issue with link to Slack and for mWeb unless the PR is related to mWeb we should not block on it given this issue

This isn't platform-specific, so I'll go ahead and finish the checklist and after you update the **Screenshots/Videos** section we are good to merge

**Comment by cead22** [comment]
I'm looking into whitelisting both your IPs you provided to prevent this issue from happening

**Comment by eVoloshchak** [comment]
> I'm looking into whitelisting both your IPs you provided to prevent this issue from happening

I don't think I have a static IP, so it won't help. And since I'm using mobile internet 60% of the time, the IP's can change quite often

**Review by eVoloshchak — DISMISSED** [review]
LGTM!

**Comment by cead22** [comment]
Ah ok

**Comment by markarmanus** [comment]
I already requested that my ip address is while listed through the form. Should potentailly help with fixing other bugs hopefully.

It seems that everything is good to go. Just waiting on someone to merge it. Thank you guys for your help !

**Comment by cead22** [comment]
@markarmanus I've whitelisted yours

**Comment by markarmanus** [comment]
@cead22 @eVoloshchak also one thing to note is tested this on web before the login issue happened. but did not record a video for it.

**Comment by markarmanus** [comment]
@cead22 Do you want me to creata a quick video for web before merging or is not necessary ?

**Comment by markarmanus** [comment]
https://user-images.githubusercontent.com/33489215/207942612-549c3645-c435-4992-9db7-ade36920a14c.mp4

**Comment by eVoloshchak** [comment]
@markarmanus, nice!
please add a separate video for each platform in the PR description following the template

**Comment by eVoloshchak** [comment]
There are some  conflicts that must be resolved, after which we can merge this

**Comment by markarmanus** [comment]
@eVoloshchak Done

**Review by eVoloshchak — DISMISSED** [review]
Looks good
cc: @cead22

**Comment by markarmanus** [comment]
@cead22 any updates ?

File: src/libs/actions/Report.js
Comment (cead22) [inline src/libs/actions/Report.js:975]:
```suggestion
    const parsedOriginalCommentHTML = parser.replace(originalCommentHTML, autolinkFilter);
```

File: src/libs/actions/Report.js
Comment (cead22) [inline src/libs/actions/Report.js:984]:
```suggestion
    if (parsedOriginalCommentHTML === htmlForNewComment.trim()) {
```

**Review by cead22 — CHANGES_REQUESTED** [review]

File: src/libs/actions/Report.js
Comment (AndrewGable) [inline src/libs/actions/Report.js:901]:
```suggestion
    const regex = /\[[^[\]]*\]\(([^()]*)\)/gm;
    const matches = [...comment.matchAll(regex)];
```

File: src/libs/actions/Report.js
Comment (AndrewGable) [inline src/libs/actions/Report.js:935]:
```suggestion
        const regex = new RegExp(`\\[([^\\[\\]]*)\\]\\(${link}\\)`, 'gm');
        const linkMatch = regex.exec(commentCopy);
```

**Review by AndrewGable — COMMENTED** [review]
Sorry @cead22 and I were discussing this PR and I noticed this. We prefer whole words instead of abbreviations.

File: src/libs/actions/Report.js
Comment (cead22) [inline src/libs/actions/Report.js:900]:
```suggestion
    const regex = /\[[^[\]]*\]\(([^()]*)\)/gm;
```
While we're fixing variable names, let's make this one `regex`, here and in `removeLinks`

**Review by cead22 — CHANGES_REQUESTED** [review]

**Comment by markarmanus** [comment]
@AndrewGable @cead22 Done

**Comment by cead22** [comment]
@markarmanus you resolved my comment about `parsedOriginalCommentHTMl` without updating the code

**Comment by cead22** [comment]
Please be sure to re-test everything after every change

**Comment by markarmanus** [comment]
@cead22 Im sorry but im not sure the comment explains what change i need to introduce, the variable name is already `parsedOriginalCommentHTML` And yes i run a quick test on web everytime.

**Comment by markarmanus** [comment]
@AndrewGable Also no need to apologize always open to improve the code.

File: src/libs/actions/Report.js
Comment (AndrewGable) [inline src/libs/actions/Report.js:975]: The last character here is still wrong, it's a lowercase `l` and should be an uppercase `L`

Comment (markarmanus) [reply src/libs/actions/Report.js]: oh ops sorry i just noticed it and fixed :D

**Review by AndrewGable — APPROVED** [review]

**Comment by markarmanus** [comment]
@eVoloshchak

**Review by cead22 — APPROVED** [review]

**Pull request merged**
